packages/ckeditor5-code-block/src/outdentcodeblockcommand.ts

```diff
@@ -179,6 +179,19 @@ function getCodeLineTextNodeAtPosition( position: Position ): Text | null {
 		nodeAtPosition = position.nodeBefore;
 	}
 
+	// <codeBlock>foo <inlineElement></inlineElement>^</codeBlock>
+	// <codeBlock><inlineElement></inlineElement>^</codeBlock>
```

Comment:
This is more complicated than it looks and I didn't understand it correctly at the first glance (and second glance). So I thought that it will be good to write a broader comment here. But then I thought that this whole function could be generalized and the first and the last `if` should not be needed.

So, let's generalize this function to: for given position, let's find the closest position that is at the beginning of a line, and then let's see if there is a text node after this position. I think this is what we want to do here regardless of the actual content (text, formatted text - I know it's not allowed, inline elements).

Beginning of the line is of course what you have in your while condition.

So I think your while loop (maybe with some tweaks) should also cover the first and the last `if` in this function. With a proper comment it may be even more understandable than the current comments. Like, there's no description for this function now and you need to check how it is used. So I think simply writing at the beginning something like I wrote above:

```
// For given position, find the closest position that is at the beginning of a line.
// Beginning of a line is after a `softBreak` element or at the beginning of a code block.
// Then, return a text node that is at the beginning of the line.
//
// <codeBlock>foobar^</codeBlock>                ->   <codeBlock>^[foobar]</codeBlock>
// <codeBlock>foo<softBreak />bar^</codeBlock>   ->   <codeBlock>foo<softBreak />^[bar]</codeBlock>
```

And a few more interesting examples. They probably won't fit in one line so you can go with:

```
// AAA
// AAA
//
// BBB
// BBB
//
// CCC
// CCC
```

Reply:
One more thing. I think it is important to write somewhere here that even though `codeBlock` disallows inline elements, we still consider them because some features may overwrite this disallow rule and put inline objects into code block anyway (e.g. merge fields).

Reply:
One more thing :smile:. How does this work for indent? I see you didn't make any changes in indent command but I am not sure if you checked scenarios where you have non-collapsed selection. In such cases you can still indent while having selection in the middle of a line so if there's similar algorithm for indent it could break too?

Also please check if multiline selections work well for indent and outdent.